lib/engine-jit/src/engine.rs

```diff
@@ -193,130 +192,117 @@ impl JITEngineInner {
+            let ptr = allocated_functions
+                .drain(0..1)
+                .map(|slice| FunctionBodyPtr(slice as *mut [_]))
+                .collect::<Vec<_>>()[0];
+            allocated_function_call_trampolines.push(ptr);
+            let trampoline = unsafe {
+                std::mem::transmute::<*const VMFunctionBody, VMTrampoline>((**ptr).as_ptr())
+            };
             self.function_call_trampolines.insert(index, trampoline);
         }
 
-        let allocated_dynamic_function_trampolines = dynamic_function_trampolines
-            .values()
-            .map(|compiled_function| {
-                let ptr = self
-                    .code_memory
-                    .allocate_for_function(registry, &compiled_function)
-                    .map_err(|message| {
-                        CompileError::Resource(format!(
-                            "failed to allocate memory for dynamic function trampolines: {}",
-                            message
-                        ))
-                    })?;
-                Ok(FunctionBodyPtr(ptr as _))
+        let allocated_dynamic_function_trampolines = allocated_functions
+            .drain(..)
+            .map(|slice| FunctionBodyPtr(slice as *mut [_]))
+            .collect::<PrimaryMap<FunctionIndex, _>>();
+
+        let mut exec_iter = allocated_executable_sections.iter();
+        let mut data_iter = allocated_data_sections.iter();
+        let allocated_custom_sections = custom_sections
+            .iter()
+            .map(|(_, section)| {
+                SectionBodyPtr(
+                    if section.protection == CustomSectionProtection::ReadExecute {
+                        exec_iter.next()
+                    } else {
+                        data_iter.next()
+                    }
+                    .unwrap()
+                    .as_ptr(),
+                )
             })
-            .collect::<Result<PrimaryMap<FunctionIndex, _>, CompileError>>()?;
+            .collect::<PrimaryMap<SectionIndex, _>>();
 
         Ok((
-            allocated_functions,
+            allocated_functions_result,
             allocated_function_call_trampolines,
             allocated_dynamic_function_trampolines,
+            allocated_custom_sections,
         ))
     }
 
     /// Make memory containing compiled code executable.
     pub(crate) fn publish_compiled_code(&mut self) {
-        self.code_memory.publish();
+        self.code_memory.last_mut().unwrap().publish();
     }
 
-    /// Publish the unwind registry into code memory.
-    pub(crate) fn publish_unwind_registry(&mut self, unwind_registry: Arc<UnwindRegistry>) {
-        self.code_memory.publish_unwind_registry(unwind_registry);
+    /// Register DWARF-type exception handling information associated with the code.
+    pub(crate) fn publish_eh_frame(&mut self, eh_frame: Option<&[u8]>) -> Result<(), CompileError> {
+        self.code_memory
+            .last_mut()
+            .unwrap()
```

Comment:
Is this guaranteed?

Reply:
Roughly these types have a) init empty b) fill with data c) publish data. If you skip fill then the publish will fail and that's fine to panic! because it's definitely an internal error inside engine-jit. The only thing I'd change here is to add an assert! that self.code_memory isn't empty, and that isn't really a change to the behaviour.

The only other issue is why the last one, as in, can you do init+fill+init+fill+publish+publish? In a word, no, we always do these actions together while setting up a new artifact. Artifact::from_parts is what drives this.

Reply:
Yeah that's fine!  Only point I'd mention is that it'd be good to have the preconditions listed in the doc comments but that sounds reasonable!